In [1]:
import warnings
warnings.filterwarnings("ignore")
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd
import os
import env
import wrangle as w

In [2]:
w.convert_txt_data()

date      time  \
0       2018-01-26  09:55:03   
1       2018-01-26  09:56:02   
2       2018-01-26  09:56:05   
3       2018-01-26  09:56:06   
4       2018-01-26  09:56:24   
...            ...       ...   
900218  2021-04-21  16:41:51   
900219  2021-04-21  16:42:02   
900220  2021-04-21  16:42:09   
900221  2021-04-21  16:44:37   
900222  2021-04-21  16:44:39   

                                                    page  user_id  cohort_id  \
0                                                      /        1        8.0   
1                                                java-ii        1        8.0   
2                    java-ii/object-oriented-programming        1        8.0   
3                     slides/object_oriented_programming        1        8.0   
4                              javascript-i/conditionals        2       22.0   
...                                                  ...      ...        ...   
900218                              jquery/personal-site       64       28.0   
900219                                 jquery/mapbox-api       64       28.0   
900220                           jquery/ajax/weather-map       64       28.0   
900221  anomaly-detection/discrete-probabilistic-methods      744       28.0   
900222                                 jquery/mapbox-api       64       28.0   

                   ip  
0        97.105.19.61  
1        97.105.19.61  
2        97.105.19.61  
3        97.105.19.61  
4        97.105.19.61  
...               ...  
900218  71.150.217.33  
900219  71.150.217.33  
900220  71.150.217.33  
900221  24.160.137.86  
900222  71.150.217.33  

[900223 rows x 6 columns]

In [3]:
w.get_cohort_data()

cohort_id                                    name  start_date    end_date  \
0           1                                  Arches  2014-02-04  2014-04-22   
1           2                                Badlands  2014-06-04  2014-08-22   
2           3                                Carlsbad  2014-09-04  2014-11-05   
3           4                                  Denali  2014-10-20  2015-01-18   
4           5                              Everglades  2014-11-18  2015-02-24   
5           6                                Franklin  2015-02-03  2015-05-26   
6           7                                 Glacier  2015-06-05  2015-10-06   
7           8                                 Hampton  2015-09-22  2016-02-06   
8           9                                  Apollo  2015-03-30  2015-07-29   
9          10                                  Balboa  2015-11-03  2016-03-11   
10         11                                     Ike  2016-01-20  2016-05-12   
11         12                                  Joshua  2016-03-08  2016-06-30   
12         13                                   Kings  2016-05-23  2016-09-15   
13         14                                  Lassen  2016-07-18  2016-11-10   
14         15                                 Mammoth  2016-09-26  2017-02-02   
15         16                                 Niagara  2016-10-26  2017-03-09   
16         17                                 Olympic  2017-02-06  2017-05-25   
17         18                               Pinnacles  2017-03-27  2017-07-20   
18         19                                  Quincy  2017-06-05  2017-09-22   
19         20                                 Redwood  2017-07-31  2017-12-06   
20         21                                 Sequoia  2017-09-27  2018-02-15   
21         22                                   Teddy  2018-01-08  2018-05-17   
22         23                                 Ulysses  2018-03-05  2018-07-19   
23         24                               Voyageurs  2018-05-29  2018-10-11   
24         25                                Wrangell  2018-07-23  2018-11-29   
25         26                                  Xanadu  2018-09-17  2019-02-08   
26         27                                Yosemite  2018-11-05  2019-04-03   
27         28                                   Staff  2014-02-04  2014-02-04   
28         29                                    Zion  2019-01-22  2019-06-04   
29         30                                     Ada  2019-02-04  2019-06-16   
30         31                               Andromeda  2019-03-18  2019-07-30   
31         32                              Betelgeuse  2019-05-28  2019-10-08   
32         33                                   Ceres  2019-07-15  2019-12-11   
33         34                                   Bayes  2019-08-19  2020-01-30   
34         51                                  Deimos  2019-09-16  2020-02-27   
35         52                                  Europa  2019-11-04  2020-04-17   
36         53                                 Fortuna  2020-01-13  2020-06-23   
37         55                                   Curie  2020-02-03  2020-07-07   
38         56                                    Apex  2020-02-24  2020-07-29   
39         57                                Ganymede  2020-03-23  2020-08-20   
40         58                                Hyperion  2020-05-26  2020-11-10   
41         59                                  Darden  2020-07-13  2021-01-12   
42         60                                      Io  2020-07-20  2021-01-21   
43         61                                    Bash  2020-07-20  2021-01-21   
44         62                                 Jupiter  2020-09-21  2021-03-30   
45        132                                 Kalypso  2020-11-02  2021-05-04   
46        133                                  Easley  2020-12-07  2021-06-08   
47        134                                    Luna  2020-12-07  2021-06-08   
48        135                      

In [4]:
df=w.acquire_anonymized_curriculum_access_data()

In [5]:
w.add_program_type(df)

date      time  \
0       2018-01-26  09:55:03   
1       2018-01-26  09:56:02   
2       2018-01-26  09:56:05   
3       2018-01-26  09:56:06   
4       2018-01-26  09:56:24   
...            ...       ...   
900218  2021-04-21  16:41:51   
900219  2021-04-21  16:42:02   
900220  2021-04-21  16:42:09   
900221  2021-04-21  16:44:37   
900222  2021-04-21  16:44:39   

                                                    page  user_id  cohort_id  \
0                                                      /        1        8.0   
1                                                java-ii        1        8.0   
2                    java-ii/object-oriented-programming        1        8.0   
3                     slides/object_oriented_programming        1        8.0   
4                              javascript-i/conditionals        2       22.0   
...                                                  ...      ...        ...   
900218                              jquery/personal-site       64       28.0   
900219                                 jquery/mapbox-api       64       28.0   
900220                           jquery/ajax/weather-map       64       28.0   
900221  anomaly-detection/discrete-probabilistic-methods      744       28.0   
900222                                 jquery/mapbox-api       64       28.0   

                   ip     name  start_date    end_date program_id  
0        97.105.19.61  Hampton  2015-09-22  2016-02-06        PHP  
1        97.105.19.61  Hampton  2015-09-22  2016-02-06        PHP  
2        97.105.19.61  Hampton  2015-09-22  2016-02-06        PHP  
3        97.105.19.61  Hampton  2015-09-22  2016-02-06        PHP  
4        97.105.19.61    Teddy  2018-01-08  2018-05-17       Java  
...               ...      ...         ...         ...        ...  
900218  71.150.217.33    Staff  2014-02-04  2014-02-04       Java  
900219  71.150.217.33    Staff  2014-02-04  2014-02-04       Java  
900220  71.150.217.33    Staff  2014-02-04  2014-02-04       Java  
900221  24.160.137.86    Staff  2014-02-04  2014-02-04       Java  
900222  71.150.217.33    Staff  2014-02-04  2014-02-04       Java  

[900223 rows x 10 columns]

In [6]:
w.convert_datetimes(df)

date      time  \
date                              
2018-01-26 2018-01-26  09:55:03   
2018-01-26 2018-01-26  09:56:02   
2018-01-26 2018-01-26  09:56:05   
2018-01-26 2018-01-26  09:56:06   
2018-01-26 2018-01-26  09:56:24   
...               ...       ...   
2021-04-21 2021-04-21  16:41:51   
2021-04-21 2021-04-21  16:42:02   
2021-04-21 2021-04-21  16:42:09   
2021-04-21 2021-04-21  16:44:37   
2021-04-21 2021-04-21  16:44:39   

                                                        page  user_id  \
date                                                                    
2018-01-26                                                 /        1   
2018-01-26                                           java-ii        1   
2018-01-26               java-ii/object-oriented-programming        1   
2018-01-26                slides/object_oriented_programming        1   
2018-01-26                         javascript-i/conditionals        2   
...                                                      ...      ...   
2021-04-21                              jquery/personal-site       64   
2021-04-21                                 jquery/mapbox-api       64   
2021-04-21                           jquery/ajax/weather-map       64   
2021-04-21  anomaly-detection/discrete-probabilistic-methods      744   
2021-04-21                                 jquery/mapbox-api       64   

            cohort_id             ip     name start_date   end_date program_id  
date                                                                            
2018-01-26        8.0   97.105.19.61  Hampton 2015-09-22 2016-02-06        PHP  
2018-01-26        8.0   97.105.19.61  Hampton 2015-09-22 2016-02-06        PHP  
2018-01-26        8.0   97.105.19.61  Hampton 2015-09-22 2016-02-06        PHP  
2018-01-26        8.0   97.105.19.61  Hampton 2015-09-22 2016-02-06        PHP  
2018-01-26       22.0   97.105.19.61    Teddy 2018-01-08 2018-05-17       Java  
...               ...            ...      ...        ...        ...        ...  
2021-04-21       28.0  71.150.217.33    Staff 2014-02-04 2014-02-04       Java  
2021-04-21       28.0  71.150.217.33    Staff 2014-02-04 2014-02-04       Java  
2021-04-21       28.0  71.150.217.33    Staff 2014-02-04 2014-02-04       Java  
2021-04-21       28.0  24.160.137.86    Staff 2014-02-04 2014-02-04       Java  
2021-04-21       28.0  71.150.217.33    Staff 2014-02-04 2014-02-04       Java  

[900223 rows x 10 columns]

In [7]:
w.clean_the_data(df)

date      time  \
date                              
2018-01-26 2018-01-26  09:55:03   
2018-01-26 2018-01-26  09:56:02   
2018-01-26 2018-01-26  09:56:05   
2018-01-26 2018-01-26  09:56:06   
2018-01-26 2018-01-26  09:56:24   
...               ...       ...   
2021-04-21 2021-04-21  16:41:51   
2021-04-21 2021-04-21  16:42:02   
2021-04-21 2021-04-21  16:42:09   
2021-04-21 2021-04-21  16:44:37   
2021-04-21 2021-04-21  16:44:39   

                                                        page  user_id  \
date                                                                    
2018-01-26                                                 /        1   
2018-01-26                                           java-ii        1   
2018-01-26               java-ii/object-oriented-programming        1   
2018-01-26                slides/object_oriented_programming        1   
2018-01-26                         javascript-i/conditionals        2   
...                                                      ...      ...   
2021-04-21                              jquery/personal-site       64   
2021-04-21                                 jquery/mapbox-api       64   
2021-04-21                           jquery/ajax/weather-map       64   
2021-04-21  anomaly-detection/discrete-probabilistic-methods      744   
2021-04-21                                 jquery/mapbox-api       64   

            cohort_id             ip     name start_date   end_date program_id  
date                                                                            
2018-01-26        8.0   97.105.19.61  Hampton 2015-09-22 2016-02-06        PHP  
2018-01-26        8.0   97.105.19.61  Hampton 2015-09-22 2016-02-06        PHP  
2018-01-26        8.0   97.105.19.61  Hampton 2015-09-22 2016-02-06        PHP  
2018-01-26        8.0   97.105.19.61  Hampton 2015-09-22 2016-02-06        PHP  
2018-01-26       22.0   97.105.19.61    Teddy 2018-01-08 2018-05-17       Java  
...               ...            ...      ...        ...        ...        ...  
2021-04-21       28.0  71.150.217.33    Staff 2014-02-04 2014-02-04       Java  
2021-04-21       28.0  71.150.217.33    Staff 2014-02-04 2014-02-04       Java  
2021-04-21       28.0  71.150.217.33    Staff 2014-02-04 2014-02-04       Java  
2021-04-21       28.0  24.160.137.86    Staff 2014-02-04 2014-02-04       Java  
2021-04-21       28.0  71.150.217.33    Staff 2014-02-04 2014-02-04       Java  

[900223 rows x 10 columns]

In [8]:
ds=df[df.program_id=='Data Science']

In [9]:
ds

date      time                page  user_id  cohort_id  \
326053 2019-08-20  09:39:58                   /      466       34.0   
326054 2019-08-20  09:39:59                   /      467       34.0   
326055 2019-08-20  09:39:59                   /      468       34.0   
326056 2019-08-20  09:40:02                   /      469       34.0   
326057 2019-08-20  09:40:08                   /      470       34.0   
...           ...       ...                 ...      ...        ...   
900182 2021-04-21  16:14:16  regression/project      841      133.0   
900183 2021-04-21  16:15:51  regression/project      841      133.0   
900190 2021-04-21  16:23:53                   /      143      133.0   
900191 2021-04-21  16:23:57  clustering/project      841      133.0   
900202 2021-04-21  16:32:50  clustering/project      841      133.0   

                    ip    name start_date   end_date    program_id  
326053    97.105.19.58   Bayes 2019-08-19 2020-01-30  Data Science  
326054    97.105.19.58   Bayes 2019-08-19 2020-01-30  Data Science  
326055    97.105.19.58   Bayes 2019-08-19 2020-01-30  Data Science  
326056    97.105.19.58   Bayes 2019-08-19 2020-01-30  Data Science  
326057    97.105.19.58   Bayes 2019-08-19 2020-01-30  Data Science  
...                ...     ...        ...        ...           ...  
900182  99.162.244.233  Easley 2020-12-07 2021-06-08  Data Science  
900183  99.162.244.233  Easley 2020-12-07 2021-06-08  Data Science  
900190  173.174.194.60  Easley 2020-12-07 2021-06-08  Data Science  
900191  99.162.244.233  Easley 2020-12-07 2021-06-08  Data Science  
900202  99.162.244.233  Easley 2020-12-07 2021-06-08  Data Science  

[103412 rows x 10 columns]

In [10]:
fe=df[df.program_id=='Front End']

In [11]:
fe

date      time                                         page  \
31627 2018-03-22  19:01:49                                            /   
31628 2018-03-22  19:01:54                             content/html-css   
31629 2018-03-22  19:01:54  content/html-css/gitbook/images/favicon.ico   
31630 2018-03-22  19:02:47                             content/html-css   
31631 2018-03-22  19:02:52           content/html-css/introduction.html   

       user_id  cohort_id             ip    name start_date   end_date  \
31627      152        9.0  207.68.209.17  Apollo 2015-03-30 2015-07-29   
31628      152        9.0  207.68.209.17  Apollo 2015-03-30 2015-07-29   
31629      152        9.0  207.68.209.17  Apollo 2015-03-30 2015-07-29   
31630      152        9.0  207.68.209.17  Apollo 2015-03-30 2015-07-29   
31631      152        9.0  207.68.209.17  Apollo 2015-03-30 2015-07-29   

      program_id  
31627  Front End  
31628  Front End  
31629  Front End  
31630  Front End  
31631  Front End

In [12]:
jv=df[df.program_id=='Java']

In [13]:
jv

date      time                                              page  \
4      2018-01-26  09:56:24                         javascript-i/conditionals   
5      2018-01-26  09:56:41                                javascript-i/loops   
6      2018-01-26  09:56:46                         javascript-i/conditionals   
7      2018-01-26  09:56:48                            javascript-i/functions   
8      2018-01-26  09:56:59                                javascript-i/loops   
...           ...       ...                                               ...   
900218 2021-04-21  16:41:51                              jquery/personal-site   
900219 2021-04-21  16:42:02                                 jquery/mapbox-api   
900220 2021-04-21  16:42:09                           jquery/ajax/weather-map   
900221 2021-04-21  16:44:37  anomaly-detection/discrete-probabilistic-methods   
900222 2021-04-21  16:44:39                                 jquery/mapbox-api   

        user_id  cohort_id             ip   name start_date   end_date  \
4             2       22.0   97.105.19.61  Teddy 2018-01-08 2018-05-17   
5             2       22.0   97.105.19.61  Teddy 2018-01-08 2018-05-17   
6             3       22.0   97.105.19.61  Teddy 2018-01-08 2018-05-17   
7             3       22.0   97.105.19.61  Teddy 2018-01-08 2018-05-17   
8             2       22.0   97.105.19.61  Teddy 2018-01-08 2018-05-17   
...         ...        ...            ...    ...        ...        ...   
900218       64       28.0  71.150.217.33  Staff 2014-02-04 2014-02-04   
900219       64       28.0  71.150.217.33  Staff 2014-02-04 2014-02-04   
900220       64       28.0  71.150.217.33  Staff 2014-02-04 2014-02-04   
900221      744       28.0  24.160.137.86  Staff 2014-02-04 2014-02-04   
900222       64       28.0  71.150.217.33  Staff 2014-02-04 2014-02-04   

       program_id  
4            Java  
5            Java  
6            Java  
7            Java  
8            Java  
...           ...  
900218       Java  
900219       Java  
900220       Java  
900221       Java  
900222       Java  

[713365 rows x 10 columns]

In [14]:
php=df[df.program_id=='PHP']

In [15]:
php

date      time                                           page  \
0      2018-01-26  09:55:03                                              /   
1      2018-01-26  09:56:02                                        java-ii   
2      2018-01-26  09:56:05            java-ii/object-oriented-programming   
3      2018-01-26  09:56:06             slides/object_oriented_programming   
30     2018-01-26  10:14:47                                              /   
...           ...       ...                                            ...   
883786 2021-04-10  00:20:55           content/javascript/conditionals.html   
883787 2021-04-10  00:21:15                  content/javascript/loops.html   
883788 2021-04-10  00:28:10   content/javascript/javascript-with-html.html   
883789 2021-04-10  00:28:11           content/javascript/conditionals.html   
883790 2021-04-10  00:28:11  content/javascript/gitbook/images/favicon.ico   

        user_id  cohort_id              ip     name start_date   end_date  \
0             1        8.0    97.105.19.61  Hampton 2015-09-22 2016-02-06   
1             1        8.0    97.105.19.61  Hampton 2015-09-22 2016-02-06   
2             1        8.0    97.105.19.61  Hampton 2015-09-22 2016-02-06   
3             1        8.0    97.105.19.61  Hampton 2015-09-22 2016-02-06   
30           11        1.0    97.105.19.61   Arches 2014-02-04 2014-04-22   
...         ...        ...             ...      ...        ...        ...   
883786       51       13.0  72.179.168.148    Kings 2016-05-23 2016-09-15   
883787       51       13.0  72.179.168.148    Kings 2016-05-23 2016-09-15   
883788       80       14.0   136.50.29.193   Lassen 2016-07-18 2016-11-10   
883789       80       14.0   136.50.29.193   Lassen 2016-07-18 2016-11-10   
883790       80       14.0   136.50.29.193   Lassen 2016-07-18 2016-11-10   

       program_id  
0             PHP  
1             PHP  
2             PHP  
3             PHP  
30            PHP  
...           ...  
883786        PHP  
883787        PHP  
883788        PHP  
883789        PHP  
883790        PHP  

[30548 rows x 10 columns]

# 1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?

In [16]:
ds.groupby(['page'])['page'].count().sort_values(ascending=False)

page
/                                           8358
search/search_index.json                    2203
classification/overview                     1785
1-fundamentals/modern-data-scientist.jpg    1655
1-fundamentals/AI-ML-DL-timeline.jpg        1651
                                            ... 
cli/4-navigating-the-filesystem                1
regression/project/null                        1
creating-charts                                1
curie-statistics-assessment                    1
%20https://github.com/RaulCPena                1
Name: page, Length: 682, dtype: int64

## Data Science accessed classification/overview 1785 times.

In [17]:
fe.groupby(['page'])['page'].count().sort_values(ascending=False)

page
content/html-css                               2
/                                              1
content/html-css/gitbook/images/favicon.ico    1
content/html-css/introduction.html             1
Name: page, dtype: int64

## Front End accessed content/html-css 2 times.

In [18]:
jv.groupby(['page'])['page'].count().sort_values(ascending=False)

page
/                                                                      35814
javascript-i                                                           17457
toc                                                                    17428
search/search_index.json                                               15212
java-iii                                                               12683
                                                                       ...  
content/mysql/relationships/sub-queries.html                               1
content/mysql/sample-database/gitbook/images/favicon.ico                   1
content/php_iii/classes-and-objects-i/constructors-destructors.html        1
content/php_iv/classes-and-objects-ii/late-static-binding.html             1
wp-login                                                                   1
Name: page, Length: 1913, dtype: int64

## Java accessed javascript-i 17457 times.

In [19]:
php.groupby(['page'])['page'].count().sort_values(ascending=False)

page
/                      1681
index.html             1011
javascript-i            736
html-css                542
spring                  501
                       ... 
interview-questions       1
13.5_Tableau              1
interview                 1
html/forms                1
4.5_imports               1
Name: page, Length: 710, dtype: int64

## PHP accessed javascript-i 736 times.

# 7. Which lessons are least accessed?

In [20]:
ds.groupby(['page'])['page'].count().sort_values()

page
%20https://github.com/RaulCPena                      1
appendix/open_data/www.census.gov                    1
appendix/open_data/www.flickr.com/services/api       1
appendix/open_data/www.openstreetmap.org             1
appendix/univariate_regression_in_excel              1
                                                  ... 
1-fundamentals/AI-ML-DL-timeline.jpg              1651
1-fundamentals/modern-data-scientist.jpg          1655
classification/overview                           1785
search/search_index.json                          2203
/                                                 8358
Name: page, Length: 682, dtype: int64

## Data Science accessed pages with the prefix appendix the least.

In [21]:
fe.groupby(['page'])['page'].count().sort_values()

page
/                                              1
content/html-css/gitbook/images/favicon.ico    1
content/html-css/introduction.html             1
content/html-css                               2
Name: page, dtype: int64

## Front End accessed pages with the prefix content/html-css the least.

In [22]:
jv.groupby(['page'])['page'].count().sort_values()

page
'                                            1
cheatsheet.tex                               1
classification/user-defined-functions        1
cli-02-listing-files                         1
cli-03-file-paths                            1
                                         ...  
java-iii                                 12683
search/search_index.json                 15212
toc                                      17428
javascript-i                             17457
/                                        35814
Name: page, Length: 1913, dtype: int64

## Java accessed pages with the prefix cli- the least.

In [23]:
php.groupby(['page'])['page'].count().sort_values()

page
html-css/media-queries                                       1
prework/cli/07-editing-files-with-vim                        1
content/appendix/javascript/functions/controllers.html       1
content/appendix/javascript/functions/models.html            1
content/appendix/javascript/functions/scope.html             1
                                                          ... 
spring                                                     501
html-css                                                   542
javascript-i                                               736
index.html                                                1011
/                                                         1681
Name: page, Length: 710, dtype: int64

## PHP accessed pages with the prefix content/appendix/javascript/functions/ the least.